tests/priority_fairness_test.go

```diff
@@ -348,3 +354,195 @@ func unfairness(vs []int) float64 {
+	s.Eventually(func() bool {
+		resp, err := s.AdminClient().GetTaskQueueTasks(ctx, &adminservice.GetTaskQueueTasksRequest{
+			Namespace:     s.Namespace().String(),
+			TaskQueue:     tv.TaskQueue().Name,
+			TaskQueueType: enumspb.TASK_QUEUE_TYPE_ACTIVITY,
+			BatchSize:     10,
+			MinPass:       1,
+		})
+		return err == nil && len(resp.GetTasks()) == 1
+	}, 10*time.Second, 100*time.Millisecond)
+
+	// Update activity options to set a new priority
+	_, err = s.FrontendClient().UpdateActivityOptions(ctx, &workflowservice.UpdateActivityOptionsRequest{
+		Namespace: s.Namespace().String(),
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: tv.WorkflowID(),
+			RunId:      startResp.GetRunId(),
+		},
+		Activity: &workflowservice.UpdateActivityOptionsRequest_Id{Id: tv.ActivityID()},
+		ActivityOptions: &activitypb.ActivityOptions{
+			Priority: updatedPriority,
+		},
+		UpdateMask: &fieldmaskpb.FieldMask{Paths: []string{"priority"}},
+	})
+	s.NoError(err)
+
+	// Query workflow to verify activity has the updated priority.
+	descResp, err = s.FrontendClient().DescribeWorkflowExecution(ctx, &workflowservice.DescribeWorkflowExecutionRequest{
+		Namespace: s.Namespace().String(),
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: tv.WorkflowID(),
+			RunId:      startResp.GetRunId(),
+		},
+	})
+	s.NoError(err)
+	s.NotNil(descResp.GetPendingActivities())
+	s.Len(descResp.GetPendingActivities(), 1)
+	s.ProtoEqual(updatedPriority, descResp.GetPendingActivities()[0].GetPriority())
+	s.ProtoEqual(updatedPriority, descResp.GetPendingActivities()[0].GetActivityOptions().GetPriority())
+
+	// Poll for activity task and verify it has the updated priority.
+	_, err = s.TaskPoller().PollAndHandleActivityTask(
+		tv,
+		func(task *workflowservice.PollActivityTaskQueueResponse) (*workflowservice.RespondActivityTaskCompletedRequest, error) {
+			s.NotNil(task)
+			s.Equal(tv.ActivityID(), task.ActivityId)
+			s.ProtoEqual(updatedPriority, task.Priority)
+			return &workflowservice.RespondActivityTaskCompletedRequest{}, nil
+		},
+		taskpoller.WithContext(ctx),
+	)
+	s.NoError(err)
+
+	// Verify that 2 activity tasks were sent to matching; and 1 was marked obsolete
+	addActivityTaskCount := 0
+	obsoleteActivityTaskCount := 0
+	snap = capture.Snapshot()
+	for _, rec := range snap[metrics.ClientRequests.Name()] {
+		for key, val := range rec.Tags {
+			if key == metrics.OperationTagName && val == metrics.MatchingClientAddActivityTaskScope {
+				addActivityTaskCount++
+			}
+		}
+	}
+	s.Equal(2, addActivityTaskCount, "Expected 2 activity tasks to be dispatched to matching")
+	for _, rec := range snap[metrics.ClientFailures.Name()] {
+		for key, val := range rec.Tags {
+			if key == metrics.ErrorTypeTagName && val == fmt.Sprintf("%T", serviceerror.ObsoleteMatchingTask{}) {
+				obsoleteActivityTaskCount++
+			}
+		}
+	}
+	s.Equal(1+obsoleteWorkflowTaskCount, obsoleteActivityTaskCount, "Expected 1 activity task to be obsolete")
+}
```

Comment:
nice test, clear and thorough. the metrics stuff doesn't add any flakiness, I hope?